### ISSUE TITLE
[ERROR] Could not parse video metadata.

### ISSUE THREAD

**Issue by muhitrhn** [DevAndromeda/youtube-sr#26]
This error appeared when trying to fetch any song whether i use url or search.
```js
0|. | Error: Could not parse video metadata!
0|. |     at Function.Util.getVideo (/home/ubuntu/node_modules/youtube-sr/build/Util.js:1:13044)
0|. |     at Function.<anonymous> (/home/ubuntu/node_modules/youtube-sr/build/main.js:1:5363)
0|. |     at step (/home/ubuntu/node_modules/youtube-sr/build/main.js:1:1854)
0|. |     at Object.next (/home/ubuntu/node_modules/youtube-sr/build/main.js:1:1127)
0|. |     at fulfilled (/home/ubuntu/node_modules/youtube-sr/build/main.js:1:528)
0|. |     at runMicrotasks (<anonymous>)
0|. |     at processTicksAndRejections (internal/process/task_queues.js:95:5)```

**Comment by Snowflake107** [comment]
This error comes from `YouTube.getVideo` not `search`. Also, looks like it has something to do with your url itself 🤔 Can you provide your code?

**Comment by muhitrhn** [comment]
> This error comes from `YouTube.getVideo` not `search`. Also, looks like it has something to do with your url itself  Can you provide your code?

Here is my play command code portion -
```js
try {
				// Attempt to create a Track from the user's video URL
				if (url.startsWith('https://www.youtube.com/playlist?list') || url.startsWith('https://m.youtube.com/playlist?list') || url.startsWith('https://youtube.com/playlist?list') || url.startsWith('https://youtu.be/playlist?list') || url.startsWith('https://www.youtu.be/playlist?list')) {
					const search = await ytsr.getPlaylist(url, {
						type: "playlist"
					})
					const tracks = await Promise.all(search.videos.map(async m => {
						return await Track.from(m.url, {
							async onStart() {
								return interaction.success("music/play:NOW_PLAYING", {
									songName: m.title
								}).catch(console.warn);
							},
							async onFinish() {
								return interaction.success("music/play:FINISHED_PLAYING", {
									songName: m.title
								}).catch(console.warn);
							},
							async onError(error) {
								console.warn(error);
								return interaction.followUp({
									content: `Error: ${error.message}`
								}).catch(console.warn);
							}
						});
					}));
					tracks.forEach(track => list.enqueue(track))
					if (tracks.length === 0) {
						interaction.success("music/play:QUEUE_ENDED");
					}
					return interaction.success("music/play:ADDED_QUEUE_COUNT", {
						songCount: tracks.length
					});
				} else if (!url.startsWith('https://www.youtube.com/') || !url.startsWith('https://m.youtube.com/') || !url.startsWith('https://youtube.com/') || !url.startsWith('https://youtu.be/') || !url.startsWith('https://www.youtu.be/')) {
					const search = await ytsr.search(url, {
						limit: 1,
						type: "video"
					});
					const vdourl = (() => {
						for (const video of search) {
							return video.url;
						}
					})();
					const track = await Track.from(vdourl, {
						async onStart() {
							return interaction.success("music/play:NOW_PLAYING", {
								songName: track.title
							}).catch(console.warn);
						},
						async onFinish() {
							return interaction.success("music/play:FINISHED_PLAYING", {
								songName: track.title
							}).catch(console.warn);
						},
						async onError(error) {
							console.warn(error);
							return interaction.followUp({
								content: `Error: ${error.message}`
							}).catch(console.warn);
						}
					});
					list.enqueue(track);
					if (track.length === 0) {
						interaction.success("music/play:QUEUE_ENDED");
					}
					return interaction.success("music/play:ADDED_QUEUE", {
						songName: track.title
					});
				} else {
					const track = await Track.from(url, {
						async onStart() {
							return interaction.success("music/play:NOW_PLAYING", {
								songName: track.title
							}).catch(console.warn);
						},
						async onFinish() {
							return interaction.success("music/play:FINISHED_PLAYING", {
								songName: track.title
							}).catch(console.warn);
						},
						async onError(error) {
							console.warn(error);
							return interaction.followUp({
								content: `Error: ${error.message}`
							}).catch(console.warn);
						}
					});
					list.enqueue(track);
					if (track.length === 0) {
						interaction.success("music/play:QUEUE_ENDED");
					}
					return interaction.success("music/play:ADDED_QUEUE", {
						songName: track.title
					});
				}
			} catch (error) {
				console.warn(error);
				return interaction.followUp(`${error}`);
			}```

**Comment by Ayeven** [comment]
I think I have the same issue `TypeError: Cannot read property 'simpleText' of undefined
    at Function.Util.getVideo (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/Util.js:1:14124)
    at Function.<anonymous> (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/main.js:1:5363)
    at step (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/main.js:1:1854)
    at Object.next (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/main.js:1:1127)
    at fulfilled (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/main.js:1:528)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)`

After investigating it further:
the said video to reproduce error :
[Harutya/Weeb song](https://www.youtube.com/watch?v=mnwj6KxAvFc). Where the subscriber property is null https://i.imgur.com/X8pGORF.png unlike some video which have `subs` property https://i.imgur.com/3EWmnBQ.png
Which function related : `<Youtube>.getVideo(url)`
Line related to the error : [channel.subscriber](https://github.com/DevSnowflake/youtube-sr/blob/11414f8cb90bd0d8782f8b02e24113a5f0d041b0/src/Util.ts#L395)
Possible fix : `subscribers: info.owner.videoOwnerRenderer.subscriberCountText.simpleText?.replace(" subscribers", "")` change to `subscribers: info.owner.videoOwnerRenderer?.subscriberCountText.simpleText ?? " ",`

p.s : the suggested fix is not tested since I don't really understand Typescript, but it should be similar to Javascript Optional Chaining which if `subscriberCountText` property is `undefined||null` it should change to empty string like you intended in the original code

What will happen after the changes?:  The user/developer will need `nodejs` version `14` at least. Though, since I already see the optional chaining in action, that won't matter much

**Comment by sujalgoel** [comment]
> > This error comes from `YouTube.getVideo` not `search`. Also, looks like it has something to do with your url itself  Can you provide your code?
> 
> Here is my play command code portion. Kinda ugly how i did the if thingy -
> 
> ```js
> try {
> 	// Attempt to create a Track from the user's video URL
> 	if (url.startsWith('https://www.youtube.com/playlist?list') || url.startsWith('https://m.youtube.com/playlist?list') || url.startsWith('https://youtube.com/playlist?list') || url.startsWith('https://youtu.be/playlist?list') || url.startsWith('https://www.youtu.be/playlist?list')) {
> 		const search = await ytsr.getPlaylist(url, {
> 			type: "playlist"
> 		})
> 		const tracks = await Promise.all(search.videos.map(async m => {
> 			return await Track.from(m.url, {
> 				async onStart() {
> 					return interaction.success("music/play:NOW_PLAYING", {
> 						songName: m.title
> 					}).catch(console.warn);
> 				},
> 				async onFinish() {
> 					return interaction.success("music/play:FINISHED_PLAYING", {
> 						songName: m.title
> 					}).catch(console.warn);
> 				},
> 				async onError(error) {
> 					console.warn(error);
> 					return interaction.followUp({
> 						content: `Error: ${error.message}`
> 					}).catch(console.warn);
> 				}
> 			});
> 		}));
> 		tracks.forEach(track => list.enqueue(track))
> 		if (tracks.length === 0) {
> 			interaction.success("music/play:QUEUE_ENDED");
> 		}
> 		return interaction.success("music/play:ADDED_QUEUE_COUNT", {
> 			songCount: tracks.length
> 		});
> 	} else if (!url.startsWith('https://www.youtube.com/') || !url.startsWith('https://m.youtube.com/') || !url.startsWith('https://youtube.com/') || !url.startsWith('https://youtu.be/') || !url.startsWith('https://www.youtu.be/')) {
> 		const search = await ytsr.search(url, {
> 			limit: 1,
> 			type: "video"
> 		});
> 		const vdourl = (() => {
> 			for (const video of search) {
> 				return video.url;
> 			}
> 		})();
> 		const track = await Track.from(vdourl, {
> 			async onStart() {
> 				return interaction.success("music/play:NOW_PLAYING", {
> 					songName: track.title
> 				}).catch(console.warn);
> 			},
> 			async onFinish() {
> 				return interaction.success("music/play:FINISHED_PLAYING", {
> 					songName: track.title
> 				}).catch(console.warn);
> 			},
> 			async onError(error) {
> 				console.warn(error);
> 				return interaction.followUp({
> 					content: `Error: ${error.message}`
> 				}).catch(console.warn);
> 			}
> 		});
> 		list.enqueue(track);
> 		if (track.length === 0) {
> 			interaction.success("music/play:QUEUE_ENDED");
> 		}
> 		return interaction.success("music/play:ADDED_QUEUE", {
> 			songName: track.title
> 		});
> 	} else {
> 		const track = await Track.from(url, {
> 			async onStart() {
> 				return interaction.success("music/play:NOW_PLAYING", {
> 					songName: track.title
> 				}).catch(console.warn);
> 			},
> 			async onFinish() {
> 				return interaction.success("music/play:FINISHED_PLAYING", {
> 					songName: track.title
> 				}).catch(console.warn);
> 			},
> 			async onError(error) {
> 				console.warn(error);
> 				return interaction.followUp({
> 					content: `Error: ${error.message}`
> 				}).catch(console.warn);
> 			}
> 		});
> 		list.enqueue(track);
> 		if (track.length === 0) {
> 			interaction.success("music/play:QUEUE_ENDED");
> 		}
> 		return interaction.success("music/play:ADDED_QUEUE", {
> 			songName: track.title
> 		});
> 	}
> } catch (error) {
> 	console.warn(error);
> 	return interaction.followUp(`${error}`);
> }
> ```

but you aren't using `.getVideo()`. So, from where this error is coming from?

**Comment by Snowflake107** [comment]
> I think I have the same issue `TypeError: Cannot read property 'simpleText' of undefined at Function.Util.getVideo (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/Util.js:1:14124) at Function.<anonymous> (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/main.js:1:5363) at step (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/main.js:1:1854) at Object.next (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/main.js:1:1127) at fulfilled (/home/ubuntu/discord-musicbot/node_modules/youtube-sr/build/main.js:1:528) at runMicrotasks (<anonymous>) at processTicksAndRejections (node:internal/process/task_queues:96:5)`
> 
> After investigating it further:
> the said video to reproduce error :
> [Harutya/Weeb song](https://www.youtube.com/watch?v=mnwj6KxAvFc). Where the subscriber property is null https://i.imgur.com/X8pGORF.png unlike some video which have `subs` property https://i.imgur.com/3EWmnBQ.png
> Which function related : `<Youtube>.getVideo(url)`
> Line related to the error : [channel.subscriber](https://github.com/DevSnowflake/youtube-sr/blob/11414f8cb90bd0d8782f8b02e24113a5f0d041b0/src/Util.ts#L395)
> Possible fix : `subscribers: info.owner.videoOwnerRenderer.subscriberCountText.simpleText?.replace(" subscribers", "")` change to `subscribers: info.owner.videoOwnerRenderer?.subscriberCountText?.simpleText ?? " ",`
> 
> p.s : the suggested fix is not tested since I don't really understand Typescript, but it should be similar to Javascript Optional Chaining which if `subscriberCountText` property is `undefined||null` it should change to empty string like you intended in the original code
> 
> What will happen after the changes?: The user/developer will need `nodejs` version `14` at least. Though, since I already see the optional chaining in action, that won't matter much

Thank you, I've fixed this in v4.1.6
Still not sure about @muhitrhn's problem though

**Comment by DevAndromeda** [comment]
stale, re-open if needed. And I'm still unable to reproduce this bug
